flow/layers/layer_state_stack.h

```diff
@@ -0,0 +1,257 @@
+   public:
+    SaveEntry() = default;
+
+    void apply(SkCanvas* canvas, DisplayListBuilder* builder) const override;
+    virtual void restore(SkCanvas* canvas,
+                         DisplayListBuilder* builder) const override;
+
+   protected:
+    virtual void do_checkerboard(SkCanvas* canvas,
+                                 DisplayListBuilder* builder) const {}
+  };
+
+  class SaveLayerEntry : public SaveEntry {
+   public:
+    SaveLayerEntry(const SkRect* bounds, bool checkerboard)
+        : bounds_(bounds ? *bounds : SkRect::MakeEmpty()),
+          has_bounds_(bounds != nullptr),
+          checkerboard_(checkerboard) {}
+
+    void apply(SkCanvas* canvas, DisplayListBuilder* builder) const override;
+
+   protected:
+    const SkRect bounds_;
+    const bool has_bounds_;
+    const bool checkerboard_;
+
+    void do_checkerboard(SkCanvas* canvas,
+                         DisplayListBuilder* builder) const override;
+
+    const SkRect* save_bounds() const {
+      return has_bounds_ ? &bounds_ : nullptr;
+    }
+  };
+
+  class OpacityEntry : public SaveLayerEntry {
+   public:
+    OpacityEntry(const SkRect* bounds, SkScalar opacity, bool checkerboard)
+        : SaveLayerEntry(bounds, checkerboard), opacity_(opacity) {}
+
+    void apply(SkCanvas* canvas, DisplayListBuilder* builder) const override;
+
+   private:
+    const SkScalar opacity_;
+  };
+
+  class ImageFilterEntry : public SaveLayerEntry {
+   public:
+    ImageFilterEntry(const SkRect* bounds,
+                     const std::shared_ptr<const DlImageFilter> filter,
+                     bool checkerboard)
+        : SaveLayerEntry(bounds, checkerboard), filter_(filter) {}
+    ~ImageFilterEntry() override = default;
+
+    void apply(SkCanvas* canvas, DisplayListBuilder* builder) const override;
+
+   private:
+    const std::shared_ptr<const DlImageFilter> filter_;
+  };
+
+  class ColorFilterEntry : public SaveLayerEntry {
+   public:
+    ColorFilterEntry(const SkRect* bounds,
+                     const std::shared_ptr<const DlColorFilter> filter,
+                     bool checkerboard)
+        : SaveLayerEntry(bounds, checkerboard), filter_(filter) {}
+    ~ColorFilterEntry() override = default;
+
+    void apply(SkCanvas* canvas, DisplayListBuilder* builder) const override;
+
+   private:
+    const std::shared_ptr<const DlColorFilter> filter_;
+  };
+
+  class BackdropFilterEntry : public SaveLayerEntry {
+   public:
+    BackdropFilterEntry(const SkRect* bounds,
+                        const std::shared_ptr<const DlImageFilter> filter,
+                        DlBlendMode blend_mode,
+                        bool checkerboard)
+        : SaveLayerEntry(bounds, checkerboard),
+          filter_(filter),
+          blend_mode_(blend_mode) {}
+    ~BackdropFilterEntry() override = default;
+
+    void apply(SkCanvas* canvas, DisplayListBuilder* builder) const override;
+
+    bool is_backdrop_filter() const override { return false; }
```

Comment:
return  true

Reply:
Eek. Cut and paste.

Reply:
And this method was eliminated in favor of a "reapply" concept that does the right thing without a test in the delegate swapping code...